### ISSUE TITLE
Julia project specific environments (Project.toml)

### ISSUE THREAD

**Issue by cscheid** [quarto-dev/quarto-cli#1483]
### Discussed in https://github.com/quarto-dev/quarto-cli/discussions/1482

<div type='discussions-op-text'>

<sup>Originally posted by **tbeason** July 21, 2022</sup>
There is some discussion on the Quarto guide about how to use various environments from R or Python. Right now, it seems that in order to use Julia code blocks I must keep any packages that I want to use in my global/default environment, rather than project specific ones. In Julia, this done by having a `Project.toml` file for the project (and perhaps a `Manifest.toml` also but not always needed).

When starting Julia you can call `julia --project=Project.toml` if it is in the current directory or you can start Julia normally and from the REPL type `]` and then `activate .`
```julia
] activate .
```
or you can write
```julia
using Pkg
Pkg.activate(".")
```

I know that the last option would work if I added it to a (each?) code block, but is there a way to just add something to the `quarto.yml` file instead to use a `Project.toml` file for the whole project?
</div>

**Comment by cscheid** [comment]
We're using IJulia for our Julia execution engine. The [IJulia guidance](https://github.com/JuliaLang/IJulia.jl/issues/928) seems to be to add

```julia
using Pkg
Pkg.activate(".")
```

to the notebook. In Quarto, it'd be sufficient to add this once, to the very first code block. (All code blocks execute in the same "context")

I agree that it would be nice for different engines to configure settings like this, but we have to design this thoughtfully to avoid a proliferation of different option styles and conventions across different engines. This is a bigger change than just the Julia setting, so I don't have a concrete timeline to offer you.

**Comment by tbeason** [comment]
If I am writing a book (in Quarto terminology), do I need to add such a command in each `.qmd` file?

**Comment by cscheid** [comment]
Yes, you would. Each `.qmd` file executes independently of the other.

**Comment by Lightup1** [comment]
In my windows machine, it just start the Jupyter kernel in the current environment folder.
But I just find that run cell in my machine will not start a Jupyter console but a REPL of julia.

**Comment by camilogarciabotero** [comment]
Just to add some other behaviors of the Julia environments:

if Julia is installed via [juliaup](https://github.com/JuliaLang/juliaup) (which will become the *de facto* language version manager), then the `Julia` resides in `~/.juliaup/bin/julia` or in `.julia/juliaup/<channel>/bin/julia` (but this is fine right away and handled by `IJulia` `Jupyer` and `Quarto`), what I wanted to point is about the `environments`. So, `Julia` store them in a accsesible dir (`.julia/environments/`) if created with the `@` inside the REPL:
 
```julia
] activate @NewEnv
``` 

So that way one can always go to this `env` in the REPL without using any path. Note that, if there is no packages added, then the env, and therefore the dir, will be deleted.

**Comment by cderv** [comment]
Posting here for reference a useful comment from Discussion related thread

----

The Julia project path may be specified by the environment variable `JULIA_PROJECT`, see https://docs.julialang.org/en/v1/manual/environment-variables/#JULIA_PROJECT.

Environment variables available at code block execution can be set in an `_environment` file, see
https://quarto.org/docs/projects/environment.html

_Originally posted by @MHellmund in https://github.com/quarto-dev/quarto-cli/discussions/1482#discussioncomment-5183728_

**Comment by MHellmund** [comment]
Let me just add: 

- The project environment where Julia starts must have the IJulia package installed.
- You may add options to the kernel.json file installed for jupyter during the installation of IJulia. E.g., I use

```{julia}
Pkg.add("IJulia")
using IJulia
installkernel("Julia", "--optimize=0", "--project=@.") 
```

`--project=@.` means: search the current directory and its parents for `Project.toml`. If nothing is found, julia takes the global environment `$HOME/.julia/environments/<version>/`

**Comment by cscheid** [comment]
We now have fairly advanced Julia project configuration with the [`julia` engine](https://quarto.org/docs/computations/julia.html#using-the-julia-engine); I'm going to go ahead and close this one.
